core/server/proxy/src/main/java/alluxio/proxy/s3/S3Handler.java

```diff
@@ -0,0 +1,518 @@
+      handler.init();
+      S3BaseTask task = null;
+      if (object != null && !object.isEmpty()) {
+        task = S3ObjectTask.Factory.create(handler);
+      } else {
+        task = S3BucketTask.Factory.create(handler);
+      }
+      handler.setS3Task(task);
+      return handler;
+    } catch (Exception ex) {
+      LOG.error("Exception during create s3handler:{}", ThreadUtils.formatStackTrace(ex));
+      throw ex;
+    }
+  }
+
+  /**
+   * Process the response returned from S3Task core logic to write to downstream.
+   * @param servletResponse
+   * @param response
+   * @throws IOException
+   */
+  public static void processResponse(HttpServletResponse servletResponse,
+                                     Response response) throws IOException {
+    try {
+      // Status
+      servletResponse.setStatus(response.getStatus());
+      // Headers
+      final MultivaluedMap<String, String> headers = response.getStringHeaders();
+      for (final Map.Entry<String, List<String>> e : headers.entrySet()) {
+        final Iterator<String> it = e.getValue().iterator();
+        if (!it.hasNext()) {
+          continue;
+        }
+        final String header = e.getKey();
+        if (servletResponse.containsHeader(header)) {
+          // replace any headers previously set with values from Jersey container response.
+          servletResponse.setHeader(header, it.next());
+        }
+        while (it.hasNext()) {
+          servletResponse.addHeader(header, it.next());
+        }
+      }
+      // Entity
+      if (response.hasEntity()) {
+        ServletOutputStream servletOut = servletResponse.getOutputStream();
+        Object entity = response.getEntity();
+        if (entity instanceof InputStream) {
+          InputStream is = (InputStream) entity;
+          byte[] bytesArray = TLS_BYTES.get();
+          int read;
+          do {
+            try {
+              read = is.read(bytesArray);
+            } catch (IOException ex) {
+              /* Alluxio thrown IOException, remapping the exception
+              and send new response to downstream again */
+              Response errorResponse = S3ErrorResponse.createErrorResponse(ex, "");
+              S3Handler.processResponse(servletResponse, errorResponse);
+              return;
+            }
+            if (read == -1) {
+              break;
+            }
+            servletOut.write(bytesArray, 0, read);
+          } while (true);
+        } else {
+          String contentStr = entity.toString();
+          int contentLen = contentStr.length();
+          servletResponse.setContentLength(contentLen);
+          servletOut.write(contentStr.getBytes());
+        }
+      }
+    } finally {
+      response.close();
+    }
+  }
+
+  /**
+   * Initialize the S3Handler object in preparation for handling the request.
+   * @throws Exception
+   */
+  public void init() throws Exception {
+    // Do Authentication of the request.
+    doAuthorization();
```

Comment:
I think it's a mislead from S3. Absolutely, the header is "Authorization", but the thing we do here is Authentication. I'd like the function name is the same as the previous note line "Do Authentication".

Reply:
done